**Demonstrating some basic pseudorandom generators.**

**MIDDLE SQUARES METHOD**

This pRNG involves the following steps:

1. Let $x$ = seed, where seed is some value of $n$ digits
2. Assign $x$ = middle $n$ digits of $c+x^2$, where $c$ is some constant
3. Save x to the list of generated pseudorandom numbers $R$
4. Go to step 2

In [77]:
# Simple generator
def middleSquares(seed, upper, size):
    R = [seed] # List of generated numbers
    size //= 2
    """
    'size' is the size of the central i.e. middle substring
    that we extract from the numerical strings of our generated numbers.
    Doing size = size // 2 above helps define the middle index interval
    from which we extract the middle of the string.
    For example, if size is given as 5, size = size // 2 => size is 2.
    Hence, our middle index interval becomes [mid-2, mid+2], where
    mid is the middle index of the string.
    """
    
    x, i = seed, 1
    while i < upper:
        x = str(123 + x**2) # Getting the numerical string
        mid = len(x) // 2 # Getting the middle index of the numerical string
        x = int(x[mid-size:mid+size+1]) # Extracting the middle of the string
        """
        The middle index interval from which we extract is
        [mid-size, mid+size])
        """
        
        R.append(x) # Appending newly generated number in the list
        i += 1
    return R

# Generator that stops when sequence starts repeating (no given upper bound)
def middleSquaresDistinct(seed, size):
    R = {seed: 0} # Dictionary of generated (key is generated number, value is index)
    size //= 2
    x, i = seed, 1
    while True:
        x = str(123 + x**2)
        mid = len(x) // 2
        x = int(x[mid-size:mid+size+1])
        try:
            x = R[x]
            """
            If we can obtain a value for the key 'x'
            i.e. the currently generated pseudorandom
            number, that means the number already exists
            in the dictionary as a key, hence
            no ValueError exception will be thrown.
            Hence, we end the function and return the dictionary.
            """
            
            return {"distinctValues": list(R.keys()),
                    "repeatStart": x,
                    "period": i - x}
        except: R[x] = i
        i += 1

Example and comparison between the above two functions...

In [83]:
A = middleSquares(52, 13, 3)
B = middleSquaresDistinct(52, 3)

print("SIMPLE MIDDLE SQUARES METHOD RESULTS\n")
print(A)
print("------------")
print("MIDDLE SQUARES METHOD RESULTS WITH DISTINCTNESS\n")
for i in B.items():
    print(i[0])
    print(i[1])
    print()

SIMPLE MIDDLE SQUARES METHOD RESULTS

[52, 827, 405, 414, 151, 292, 538, 956, 405, 414, 151, 292, 538]
------------
MIDDLE SQUARES METHOD RESULTS WITH DISTINCTNESS

distinctValues
[52, 827, 405, 414, 151, 292, 538, 956]

repeatStart
2

period
6



Hence, we see that repeating sequence starts from the 7th index i.e. 8th number in the pseudorandom sequence, and the repeating sequence has 8 distinct numbers (52, 827, 405, 414, 151, 292, 538, 956). Below, we can observe the number of distinct pseudorandom numbers generated (before repeating sequence) for different seeds and middle interval sizes...

In [79]:
exampleSeeds = [3135, 31316, 141241, 31222342, 211411201]
exampleMidSizes = [3, 6, 6, 7, 10]

# Examples (in a loop)
for i in range(0, len(exampleSeeds)):
    R = middleSquaresDistinct(exampleSeeds[i], exampleMidSizes[i])
    print("\nExample", i+1)
    print("Sample values:", R['distinctValues'][0:4])
    print("Repeat start:", R['repeatStart'])
    print("Period:", R['period'])


Example 1
Sample values: [3135, 283, 21, 564]
Repeat start: 32
Period: 6

Example 2
Sample values: [31316, 8069197, 1940224, 4469170]
Repeat start: 81
Period: 660

Example 3
Sample values: [141241, 9490202, 3934000, 6356000]
Repeat start: 1080
Period: 1871

Example 4
Sample values: [31222342, 3463996, 9268288, 1162451]
Repeat start: 569
Period: 1871

Example 5
Sample values: [211411201, 94695908262, 50415651198, 78857183988]
Repeat start: 83909
Period: 2


**LINEAR CONGRUENTIAL GENERATOR**

The terms of this pseudorandom sequence are generated using $x_{n+1}=(ax_n+b)\mod M$, where $x_{n+1}$ is the next term, $x_n$ is th previous term, $a$, $b$ and $M$ are constants. Note that $x_0$ is the seed of this pseudorandom sequence.

In [81]:
def lcg(seed, a, b, M):
    R = {seed: 0}
    x, i = seed, 1
    while True:
        x = (a*x+b) % M
        try:
            x = R[x]
            return {"distinctValues": list(R.keys()),
                    "repeatStart": x,
                    "period": i - x}
        except: R[x] = i
        i += 1

Below, we can observe the number of distinct pseudorandom numbers generated (before repeating sequence) for different seeds and middle interval sizes...

In [82]:
exampleSeeds = [31, 231, 34, 14, 52]
a = [3, 6, 6, 7, 10]
b = [31, 2, 41, 7, 13]
M = [412, 41, 18, 97, 43]

# Examples (in a loop)
for i in range(0, len(exampleSeeds)):
    R = lcg(exampleSeeds[i], a[i], b[i], M[i])
    print("\nExample", i+1)
    print("Sample values:", R['distinctValues'][0:4])
    print("Repeat start:", R['repeatStart'])
    print("Period:", R['period'])


Example 1
Sample values: [31, 124, 403, 4]
Repeat start: 0
Period: 34

Example 2
Sample values: [231, 35, 7, 3]
Repeat start: 1
Period: 40

Example 3
Sample values: [34, 11, 17]
Repeat start: 2
Period: 1

Example 4
Sample values: [14, 8, 63, 60]
Repeat start: 0
Period: 96

Example 5
Sample values: [52, 17, 11, 37]
Repeat start: 1
Period: 21


Some other similar modulus-based pRNGs are defined as follows...

- $x_{n+1}=x^k\mod M$, where $k$ and $M$ are constant positive integers
- $x_{n+1}=k^{x_n}\mod M$, where $k$ and $M$ are constant positive integers